### ISSUE TITLE
MongoDB does not start correctly on staging.hackafake.com

### ISSUE THREAD

**Issue by gmacario** [hackafake/hackafake-backend#31]
Service `mongodb` does not start correctly when hackafake-backend is deployed to root@staging.hackafake.com

```
root@cs-176-221-51-111:/var/tmp/hackafake-backend-master (master)$ docker-compose up mongodb
Starting hackafake-backend-master_mongodb_1 ... done
Attaching to hackafake-backend-master_mongodb_1
mongodb_1  | 2018-08-01T12:34:51.418+0000 I JOURNAL  [initandlisten] journal dir=/data/db/journal
mongodb_1  | 2018-08-01T12:34:51.418+0000 I JOURNAL  [initandlisten] recover : no journal files present, no recovery needed
mongodb_1  | 2018-08-01T12:34:51.418+0000 I JOURNAL  [initandlisten]
mongodb_1  | 2018-08-01T12:34:51.418+0000 E JOURNAL  [initandlisten] Insufficient free space for journal files
mongodb_1  | 2018-08-01T12:34:51.418+0000 I JOURNAL  [initandlisten] Please make at least 3379MB available in /data/db/journal or use --smallfiles
mongodb_1  | 2018-08-01T12:34:51.418+0000 I JOURNAL  [initandlisten]
mongodb_1  | 2018-08-01T12:34:51.418+0000 I STORAGE  [initandlisten] exception in initAndListen: 15926 Insufficient free space for journals, terminating
mongodb_1  | 2018-08-01T12:34:51.419+0000 I CONTROL  [initandlisten] now exiting
mongodb_1  | 2018-08-01T12:34:51.419+0000 I NETWORK  [initandlisten] shutdown: going to close listening sockets...
mongodb_1  | 2018-08-01T12:34:51.419+0000 I NETWORK  [initandlisten] removing socket file: /tmp/mongodb-27017.sock
mongodb_1  | 2018-08-01T12:34:51.419+0000 I NETWORK  [initandlisten] shutdown: going to flush diaglog...
mongodb_1  | 2018-08-01T12:34:51.419+0000 I NETWORK  [initandlisten] shutdown: going to close sockets...
mongodb_1  | 2018-08-01T12:34:51.419+0000 I STORAGE  [initandlisten] shutdown: waiting for fs preallocator...
mongodb_1  | 2018-08-01T12:34:51.419+0000 I STORAGE  [initandlisten] shutdown: final commit...
mongodb_1  | 2018-08-01T12:34:51.419+0000 I STORAGE  [initandlisten] shutdown: closing all files...
mongodb_1  | 2018-08-01T12:34:51.419+0000 I STORAGE  [initandlisten] closeAllFiles() finished
mongodb_1  | 2018-08-01T12:34:51.419+0000 I CONTROL  [initandlisten] dbexit:  rc: 100
hackafake-backend-master_mongodb_1 exited with code 100
root@cs-176-221-51-111:/var/tmp/hackafake-backend-master (master)*$
```

**Comment by gmacario** [comment]
POSSIBLE EXPLANATION: To mitigate potential security issues, docker engine runs as user "docker" (uid 999)

**Comment by gmacario** [comment]
Try deploying project on non-root user `hackafake` (uid 900)

```
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$ id
uid=900(hackafake) gid=900(hackafake) groups=900(hackafake),999(docker)
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$ ls -la
total 72
drwxr-xr-x 6 hackafake hackafake 4096 Aug  2 06:41 .
drwxrwxrwt 5 root      root      4096 Aug  2 01:15 ..
drwxr-xr-x 2 hackafake hackafake 4096 Jul  5 23:04 bot
-rw-r--r-- 1 hackafake hackafake  269 Jul  5 23:04 docker-compose.yml
-rw-r--r-- 1 hackafake hackafake  480 Jul  5 23:04 Dockerfile
-rwxr-xr-x 1 hackafake hackafake   57 Jul  5 23:04 entrypoint.sh
drwxr-xr-x 8 hackafake hackafake 4096 Aug  2 06:42 .git
-rw-r--r-- 1 hackafake hackafake 1188 Jul  5 23:04 .gitignore
-rw-r--r-- 1 hackafake hackafake 1978 Aug  1 15:00 Jenkinsfile
-rw-r--r-- 1 hackafake hackafake 1511 Jul  5 23:04 LICENSE
-rwxr-xr-x 1 hackafake hackafake  222 Jul  5 23:04 manage.py
-rw-r--r-- 1 hackafake hackafake  289 Jul  5 23:04 nginx.conf
-rw-r--r-- 1 hackafake hackafake  866 Jul 31 08:57 README.md
-rw-r--r-- 1 hackafake hackafake  707 Jul  5 23:04 requirements.txt
drwxr-xr-x 2 hackafake hackafake 4096 Jul  5 23:04 server
-rw-r--r-- 1 hackafake hackafake  551 Jul  5 23:04 supervisord.conf
drwxr-xr-x 2 hackafake hackafake 4096 Jul  5 23:04 tests
-rw-r--r-- 1 hackafake hackafake  160 Jul  5 23:04 uwsgi.ini
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$
``` 

Result: no progress

```
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$ docker-compose up
 mongodb
Creating network "hackafake-backend-master_default" with the default driver
Creating hackafake-backend-master_mongodb_1 ... done
Attaching to hackafake-backend-master_mongodb_1
mongodb_1  | 2018-08-02T04:43:39.199+0000 I JOURNAL  [initandlisten] journal dir=/data/db/journal
mongodb_1  | 2018-08-02T04:43:39.199+0000 I JOURNAL  [initandlisten] recover : no journal files present, no recovery needed
mongodb_1  | 2018-08-02T04:43:39.199+0000 I JOURNAL  [initandlisten]
mongodb_1  | 2018-08-02T04:43:39.199+0000 E JOURNAL  [initandlisten] Insufficient free space for journal files
mongodb_1  | 2018-08-02T04:43:39.199+0000 I JOURNAL  [initandlisten] Please make at least 3379MB available in /data/db/journal or use --smallfiles
mongodb_1  | 2018-08-02T04:43:39.199+0000 I JOURNAL  [initandlisten]
mongodb_1  | 2018-08-02T04:43:39.199+0000 I STORAGE  [initandlisten] exception in initAndListen: 15926 Insufficient free space for journals, terminating
mongodb_1  | 2018-08-02T04:43:39.199+0000 I CONTROL  [initandlisten] now exiting
mongodb_1  | 2018-08-02T04:43:39.200+0000 I NETWORK  [initandlisten] shutdown: going to close listening sockets...
mongodb_1  | 2018-08-02T04:43:39.200+0000 I NETWORK  [initandlisten] removing socket file: /tmp/mongodb-27017.sock
mongodb_1  | 2018-08-02T04:43:39.200+0000 I NETWORK  [initandlisten] shutdown: going to flush diaglog...
mongodb_1  | 2018-08-02T04:43:39.200+0000 I NETWORK  [initandlisten] shutdown: going to close sockets...
mongodb_1  | 2018-08-02T04:43:39.200+0000 I STORAGE  [initandlisten] shutdown: waiting for fs preallocator...
mongodb_1  | 2018-08-02T04:43:39.200+0000 I STORAGE  [initandlisten] shutdown: final commit...
mongodb_1  | 2018-08-02T04:43:39.200+0000 I STORAGE  [initandlisten] shutdown: closing all files...
mongodb_1  | 2018-08-02T04:43:39.200+0000 I STORAGE  [initandlisten] closeAllFiles() finished
mongodb_1  | 2018-08-02T04:43:39.200+0000 I CONTROL  [initandlisten] dbexit:  rc: 100
hackafake-backend-master_mongodb_1 exited with code 100
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$
```

Inspect the new files and directories created on the host

```
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$ ls -la data/
total 12
drwxr-xr-x 3       999 root      4096 Aug  2 06:43 .
drwxr-xr-x 7 hackafake hackafake 4096 Aug  2 06:43 ..
drwxr-xr-x 2       999 docker    4096 Aug  2 06:43 journal
-rwxr-xr-x 1       999 docker       0 Aug  2 06:43 mongod.lock
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$
```

**Comment by gmacario** [comment]
As a matter of fact, the root cause was clearly explained in the error message

```
mongodb_1  | 2018-08-02T04:43:39.199+0000 E JOURNAL  [initandlisten] Insufficient free space for journal files
mongodb_1  | 2018-08-02T04:43:39.199+0000 I JOURNAL  [initandlisten] Please make at least 3379MB available in /data/db/journal or use --smallfiles
```

Inspect disk space

```
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$ df -h .
Filesystem             Size  Used Avail Use% Mounted on
/dev/mapper/vg00-root  6.6G  4.7G  1.6G  75% /
hackafake@cs-176-221-51-111:/var/tmp/hackafake-backend-master$
```

**Comment by gmacario** [comment]
Suggested workaround from upstream https://github.com/dockerfile/mongodb/issues/9

> @jareware instead of modifying the docker file you can run the container this way
> `docker run -d -p 27017:27017 --name test-mongo mongo mongod --smallfiles`

**Comment by gmacario** [comment]
@ludusrusso do you see any issues in adding "--smallfiles" option when starting MongoDB?

**Comment by gmacario** [comment]
Merged PR #36 to automatically deploy hackafake-server to http://staging.hackafake.it:8080/

**Comment by gmacario** [comment]
Testing hackafake Open API via http://staging.hackafake.com:8080/#!/default/get_counter_resource, everything seems to work now.
